Document push notifications debugging (#248)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/DebuggingPushNotifications.md

# Debugging Push Notifications

To sucessfully debug Push Notification on your device make sure that you have proper configurations set up as described in this document.

> - [PubNub](#pubnub)

Reviewer: Nit  - TL;DR tends to be on the top :)

Author: Then no one will read the doc! =)

Reviewer: And they should read the doc, even if they just want TL;DR? ;) 

In the end we tend to have summary :)

Anyway, I think the document needs a clearer structure

Author: of course they should read the doc, otherwise I just wrote it for myself =P
I've added more sections as you suggested

- [TLDR](#tldr)

## PubNub

[Pub Nub](https://admin.pubnub.com/) is a service we use to deliver notifications to APNS (apple Push Notifications Service) or FCM (Firebase Cloud Messaging). If you don't have access to Pub Nub yet ask other members of the team to add your work email to our account.

> Each application that supports push notifications has a separate configuration on PubNub. Depending on your access level you may have access to all the apps (admin) or to specific apps. All the apps are grouped under `steven.hamblin` account.

Reviewer: It's worth mentioning here - who can give access to required apps

Author: AFAIK we don't have specific PubNub "owner" (or any other service owner) so I wouldn't put it under someone specific prerogative. It should be part of on boarding checklist though.

Apart from certificate we need to set `APNS Environment`. This is not related to _our_ environments (dev, preprod or prod) but as the name of this setting says to APNS itself. It is possible to set this to Development or Production. Production environment is used to deliver notifications to the apps distributed through Hockeyapp, Testflight or AppStore. Development environment is used to send notifications to the apps deployed to device locally with Xcode.

For iOS we have separate key sets for push and VoIP notifications because PubNub as of Sept 2019 didn't yet support using single certificate for both types of notifications. Key sets for push notifications are prefixed with `ios-push` and are associated with regular push notifications certificate instead of VoIP certificate. All the ceritificates for all the apps can be found in 1Password iOS vault.

## TLDR
When you are testing push notifications make sure that:

> - the app is running against exepcted backend environment

Reviewer: ```suggestion
- The app is running against expected backend environment
```

> - PubNub configuration for this environment set to Production APNS Environment

Reviewer: ```suggestion
- PubNub configuration for this environment set to correct APNS Environment

	- open the key set for this environment
	- tap on "Replace  Certificate" button
	- find the correct certificate PEM file in 1Password iOS vault and upload it

> 		- for non production environment use Enterprise certificate

Reviewer: ```suggestion
		- for non-production environment use Enterprise certificate
```

		- for production environment only use AppStore certificate (or simply just never change production key set)

> 	- select proper APNS environment:

Reviewer: ```suggestion
	- select proper APNS environment:(not to be confused with backend environment)
```

> 		- for local builds (made from Xcode) select Development environment

Reviewer: ```suggestion
		- for local builds (built from Xcode) select Development environment
```

		- for Hockeyapp builds select Production environment

> - make sure you notify in ios slack channel that you are changing PubNub environment

Reviewer: ```suggestion
- make sure you notify in ios slack channel that you are changing PubNub environment to development
```

> - remember to change this configuration back when you are done and notificy ios channel

Reviewer: ```suggestion
- REMEMBER to change APNS environment back to production when you are done and notify ios slack channel. This is so that testers can continue testing notifications in on AppCentre builds.
```

> - you are on Babylon-Partners WiFi network (West Office)

Reviewer: Why do we need to be on this Wifi? It all works from home as well for me.

Author: In the office it was reported few times that some wifi network used for some QA is not suitable. Don't know what's special about it though.

Reviewer: Then you can put, check that you are not on that QA WiFi. Otherwise, a remote person will think that push notifications are not working for him/her because he/she are not on correct WiFi

Author: ```suggestion
- you are on Babylon-Partners WiFi network (if you are in the West Office)
```

> - you are on Babylon-Partners WiFi network (West Office)

Reviewer: Another step to add is to confirm that the certificates are not expired. 

Do we have a process in place to replace certificates in production environment right before they expire?

Author: This is a manual process and PubNub already checks for expiration - it will have the section in with certificate in red. Renewing certificate is a separate matter though IMO.

Reviewer: My bad to ask about renewing certificate here. I got curious. 

I was in a situation for Telus, where PubNub did not show that the certificate had been made invalid for some reason and I had to login to developer account to learn this

# Debugging Push Notifications

To sucessfully debug Push Notification on your device make sure that you have proper configurations set up as described in this document.

- [TLDR](#tldr)
- [PubNub](#pubnub)
- [Key sets](#key-sets)
- [APNS certificates and environments](#apns-certificates-and-environments)

## PubNub

> [Pub Nub](https://admin.pubnub.com/) is a service we use to deliver notifications to APNS (apple Push Notifications Service) or FCM (Firebase Cloud Messaging). If you don't have access to Pub Nub yet ask other members of the team to add your work email to our account.

Author: ```suggestion
[Pub Nub](https://admin.pubnub.com/) is a service we use to deliver notifications to APNS (Apple Push Notifications Service) or FCM (Firebase Cloud Messaging). If you don't have access to Pub Nub yet ask other members of the team to add your work email to our account.
```

Apart from certificate we need to set `APNS Environment` in each key set. This is not related to _our_ environments (dev, preprod or prod) but as the name of this setting suggests to APNS itself. It is possible to set this to Development or Production. Production environment is used to deliver notifications to the apps distributed through Hockeyapp, Testflight or AppStore. Development environment is used to send notifications to the apps deployed to device locally with Xcode.

Example 1: to test notifications locally on pre-prod environment APNS environment for `preprod` and `preprod-ios-push` key sets should be set to `Development`
Example 2: to test notifications on Hockeyapp build on pre-prod environment APNS environment for `preprod` and `preprod-ios-push` key sets should be set to `Production`

> 

Reviewer: Love those examples. Great for noobs

## TLDR
When you are testing push notifications make sure that:

- The app is running against expected backend environment
- PubNub configuration for this environment set to correct APNS Environment
	- open the key set for this environment
	- tap on "Replace  Certificate" button
	- find the correct certificate PEM file in 1Password iOS vault and upload it
		- for non-production environment use Enterprise certificate
		- for production environment only use AppStore certificate (or simply just never change production key set)
	- select proper APNS environment:
		- for local builds (built from Xcode) select Development environment
		- for Hockeyapp builds select Production environment

> - make sure you notify in ios slack channel that you are changing PubNub environment

Reviewer: Love those easy-to-digest nuggets of bullet points. I feel they can move to the top section as well. Maybe under - `Changing the APNS environment` and explain the daft PubNub UI for it.

> - REMEMBER to change APNS environment back to production when you are done and notify ios slack channel. This is so that testers can continue testing notifications on AppCentre builds.

Author: ```suggestion
- REMEMBER to change APNS environment back to production when you are done and notify ios slack channel. This is so that testing can continue on AppCentre builds.
```

> - you are on Babylon-Partners WiFi network (if you are in the West Office)

Reviewer: (I think I left this comment before but it's lost now) 
We need to change the wording that one should make sure that they are not on the rogue QA network where push notifications don't work. This can be confusing

> - you are on Babylon-Partners WiFi network (if you are in the West Office)

Author: ```suggestion
- if you are working from the West Office make sure you are on Babylon-Partners WiFi network as BabylonQA network may have issues with receiving push notifications
```
